Configurable user IDs

etcwilde opened the pull request:
When working with these containers to debug issues locally, it's sometimes useful to have the source directory live on the host, while building in the container. In order for permissions to line up, the user ID numbers need to match or the build-user may not be able to create the necessary directories.

This patch allows setting the build-user user ID so that the permissions line up and the user in the container and host can manipulate the files.

Override the build user ID with the current user's ID is done with the following command to ensure that the permissions on any volumes are kept in sync is the following:
`docker build --build-arg BUILD_USER_ID=$(id -u) ...`

etcwilde committed:
Configurable user IDs

When working with these containers to debug issues locally, it's
sometimes useful to have the source directory live on the host, while
building in the container. In order for permissions to line up, the user
ID numbers need to match or the build-user may not be able to create
the necessary directories.

This patch allows setting the build-user user ID so that the permissions
line up and the user in the container and host can manipulate the files.

Override the build user ID with the current user's ID is done with the
following command to ensure that the permissions on any volumes are
kept in sync is the following:
`docker build --build-arg BUILD_USER_ID=$(id -u) ...`

etcwilde commented:
@swift-ci please test

shahmishal commented:
In the artifact step, you should see a link to download the logs. 
> Artifact download URL: https://github.com/swiftlang/swift-docker/actions/runs/13317949874/artifacts/2589170704

etcwilde commented:
CentOS 7: 
```
#6 0.359 14: curl#6 - "Could not resolve host: mirrorlist.centos.org; Unknown error"
```

Ubuntu 23.10:
```
#7 0.293 Ign:1 http://security.ubuntu.com/ubuntu mantic-security InRelease
#7 0.293 Ign:2 http://archive.ubuntu.com/ubuntu mantic InRelease
#7 0.369 Err:3 http://security.ubuntu.com/ubuntu mantic-security Release
#7 0.369   404  Not Found [IP: 185.125.190.83 80]
#7 0.369 Ign:4 http://archive.ubuntu.com/ubuntu mantic-updates InRelease
#7 0.445 Ign:5 http://archive.ubuntu.com/ubuntu mantic-backports InRelease
#7 0.521 Err:6 http://archive.ubuntu.com/ubuntu mantic Release
#7 0.521   404  Not Found [IP: 185.125.190.83 80]
#7 0.598 Err:7 http://archive.ubuntu.com/ubuntu mantic-updates Release
#7 0.598   404  Not Found [IP: 185.125.190.83 80]
#7 0.674 Err:8 http://archive.ubuntu.com/ubuntu mantic-backports Release
#7 0.674   404  Not Found [IP: 185.125.190.83 80]
```

Ubuntu 24.04:
```
tar: usr/lib/libLTO.so.15git: Cannot write: No space left on device
...
```

Docc:
```
#5 ERROR: failed to register layer: write /usr/lib/x86_64-linux-gnu/libicudata.so.57.1: no space left on device
```

Static SDK
```
ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref c134ce27-11c9-4288-a10a-13dadbadd510::xc4kljypnmeqd83a464gmh764: "/resources": not found
```

Most of these look like either the distro is EOL or an infra issue. I can avoid the EOL distros. I'm not sure what is happening with the static SDK though. Will see if I can reproduce that. This change shouldn't have an effect.

etcwilde commented:
Ah, `ci_test.py` is trying to build the static SDK from a different directory than the directory that the Dockerfile is in, which doesn't work because the Dockerfile copies in the resources directory, which is situated next to the Dockerfile.

`Running: docker build --no-cache=true -f swift-ci/sdks/static-linux/Dockerfile .`

https://github.com/swiftlang/swift-docker/blob/e275ce1ba3382615aa7e3b0271725ac0bb55fba5/swift-ci/sdks/static-linux/Dockerfile#L88

etcwilde committed:
Configurable user IDs

When working with these containers to debug issues locally, it's
sometimes useful to have the source directory live on the host, while
building in the container. In order for permissions to line up, the user
ID numbers need to match or the build-user may not be able to create
the necessary directories.

This patch allows setting the build-user user ID so that the permissions
line up and the user in the container and host can manipulate the files.

Override the build user ID with the current user's ID is done with the
following command to ensure that the permissions on any volumes are
kept in sync is the following:
`docker build --build-arg BUILD_USER_ID=$(id -u) --build-arg BUILD_GROUP_ID=$(id -g) ...`

Note: Not including Fedora-Rawhide. The `libblocksruntime-static`
package is no longer available on the latest versions of Fedora Rawhide.
Removing the package from the `dnf install` command will fix building
the image, but we will need to come back and verify that the image is
usable later.
